psi4/src/psi4/libfock/snLinK.cc

```diff
@@ -0,0 +1,492 @@
+        timer_on("snLinK: Transform K");
+        
+        SharedMatrix K_buffer = nullptr; 
+        if (is_spherical_basis) {
+            // need to reorder Psi4 exchange matrix to CCA ordering if in spherical harmonics
+            if (permutation_matrix_.has_value()) { 
+                auto permutation_matrix_val = permutation_matrix_.value();
+                auto K_eigen_permute = K[iD]->eigen_map();
+                K_eigen_permute = permutation_matrix_val * K_eigen_permute * permutation_matrix_val.transpose();
+            }
+            
+            // also need to transform D to cartesian coordinates if requested/required
+            if (force_cartesian_) { 
+                K_buffer = std::make_shared<Matrix>();
+                K_buffer->transform(K[iD], sph_to_cart_matrix_);
+            } else {
+                K_buffer = K[iD];
+            }
+        // natively-cartesian bases dont require the reordering/transforming above
+        } else {
+            K_buffer = K[iD];
+        }
+        auto K_buffer_eigen = K_buffer->eigen_map(); 
+        timer_off("snLinK: Transform K");
+        
+        timer_on("snLinK: Execute integrator");
+        // compute delta K if incfock iteration... 
+        if (incfock_iter_) { 
+            // K buffer (i.e., delta K) must be back-transformed and added to Psi4 K separately if cartesian transformation is forced...
+            if (force_cartesian_ && is_spherical_basis) {
+                K_buffer_eigen = integrator_->eval_exx(D_buffer_eigen, integrator_settings_);
+                K_buffer->back_transform(sph_to_cart_matrix_);
+                K[iD]->add(K_buffer);
+            // ... otherwise the computation and addition can be bundled together 
+            } else {
+                K_buffer_eigen += integrator_->eval_exx(D_buffer_eigen, integrator_settings_);
+            }
+
+        // ... else compute full K 
+        } else {
+            K_buffer_eigen = integrator_->eval_exx(D_buffer_eigen, integrator_settings_);        
+            if (force_cartesian_ && is_spherical_basis) {
+                K[iD]->back_transform(K_buffer, sph_to_cart_matrix_);          
+            }
+        }
+        timer_off("snLinK: Execute integrator");
+
+        // now we need to reverse the CCA reordering previously performed
+        timer_on("snLinK: Back-transform D and K");
+        if (permutation_matrix_.has_value()) {
+            auto permutation_matrix_val = permutation_matrix_ .value();
+            auto D_eigen_permute = D[iD]->eigen_map();
+            D_eigen_permute = permutation_matrix_val.transpose() * D_eigen_permute * permutation_matrix_val; 
+
+            auto K_eigen_permute = K[iD]->eigen_map();
+            K_eigen_permute = permutation_matrix_val.transpose() * K_eigen_permute * permutation_matrix_val; 
+        }
+        timer_off("snLinK: Back-transform D and K");
+      
+        // symmetrize K if applicable
+        if (lr_symmetric_) {
```

Comment:
Is this needed for correctness?

Reply:
It is. A while back, I noticed that when I removed this part of the code, the resulting matrix is very slightly non-symmetric (~E-14 norm for `K - K^T`). Theloss of symetry presumably happens during the back-transforms, perhaps due to some numerical roundoff errors.

Reply:
Does a difference O(1E-14) yield measurable impact on the SCF? It shouldn't, as any of the errors you're getting are just an accumulation of things O(EPS), and you're just going to be looking at one of the triangles in the eigensolver.

The reason I ask this is that these kinds of operations don't scale very well - it's likely this would become a bottleneck for big systems.

Reply:
I decided to double check this, since it was a while ago that I first saw the "non-symmetry". On double-check, the RMS of `K - K^T` is actually even smaller than what I thought - `E-18` for the system I tested. Since this is below machine precision, it may mean that there isn't actually any real non-symmetry present. So I'll get rid of the explicit symmetrization for now.